dev/iterator.h

```diff
@@ -16,45 +16,49 @@
 #include "util.h"
 
 namespace sqlite_orm {
-
     namespace internal {
 
-        template<class V>
+        template<class O, class DBOs>
         struct iterator_t {
-            using view_type = V;
-            using value_type = typename view_type::mapped_type;
+            using value_type = O;
+            using db_objects_type = DBOs;
+
+          private:
+            /**
+                pointer to the view's db objects member variable.
+                only null for the default constructed iterator.
+             */
+            const db_objects_type** db_objects = nullptr;
 
-          protected:
             /**
              *  shared_ptr is used over unique_ptr here
              *  so that the iterator can be copyable.
              */
             std::shared_ptr<sqlite3_stmt> stmt;
 
-            // only null for the default constructed iterator
-            view_type* view = nullptr;
-
             /**
              *  shared_ptr is used over unique_ptr here
              *  so that the iterator can be copyable.
              */
             std::shared_ptr<value_type> current;
 
-            void extract_value() {
-                auto& dbObjects = obtain_db_objects(this->view->storage);
+            void extract_object() {
                 this->current = std::make_shared<value_type>();
                 object_from_column_builder<value_type> builder{*this->current, this->stmt.get()};
-                pick_table<value_type>(dbObjects).for_each_column(builder);
+                assert(*this->db_objects);
```

Comment:
`assert` performs nothing in release build. Let's replace `assert` with exception of something else which works in both debug and release build.

Reply:
That's my intention. The assert is just there to help in debugging of something you are never supposed to do in the first place (i.e. using an invalid iterator), and in release builds it would dereference a nullptr, which is good, and much better than the dangling pointer that we had before.

Reply:
> in release builds it would dereference a nullptr

that leads to UB, let's avoid spawning UB please. Originally there are no places in the lib which perform dereferencing before changing for `nullptr` to reduce UB which is the reason why C++ is hated a lot. Libs which spawn UB a lot are hated more

Reply:
It's not UB - it's a well-defined null pointer exception (because `~view_t()` sets it to a `nullptr`.
If you compare the diff, the iterator had a pointer to the view. If the view went out of scope and an iterator would have been accessed, then you hit UB. So contrary to what you say, there is a place that performs unchecked dereferencing. And it was never checked.

I stand by my argument that this is much worse than a controlled and well-defined null pointer exception, which - again - should not occur.
It's a programming error of the programmer still using an iterator after the lifetime of a view. That's BTW no different than other iterators in general.

Reply:
But I have a completely different suggestion: we don't need the check at all. The iterator exists perfectly without the view, it is self-sufficient, so to speak, and the ‘storage’ together with the ‘db objects’ is a global instance anyway.

Reply:
simple dereferencing a null pointer never leads to exception, it produces a reference which equals `nullptr` which is UB. Try it by yourself.
```c++
int *pointer = nullptr;
int &reference = *pointer;  // no exception is thrown here
```

Reply:
It will be thrown once you access `reference`.
![image](https://github.com/fnc12/sqlite_orm/assets/5430212/29d78dbb-49eb-4472-8451-8b75a9048b92)

Reply:
yes but it is not guaranteed by C++ standard. C++ standard only guarantees UB in that case, but segfault or any exception is compiler specific thing. This is what I am talking about. Of course in 99% cases compilers will fire segfault in case of memory violation but it is result of UB which I am talking about. If we can skip UB and prevent it with regular C++ exception which can be caught in contrast to segfault better prevent it

Reply:
> But I have a completely different suggestion: we don't need the check at all. The iterator exists perfectly without the view, it is self-sufficient, so to speak, and the ‘storage’ together with the ‘db objects’ is a global instance anyway.

Please apply your suggestion - it sounds good